include/sys/onoff.h

```diff
@@ -0,0 +1,533 @@
+/*
+ * Copyright (c) 2019 Peter Bigot Consulting, LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_ONOFF_H_
+#define ZEPHYR_INCLUDE_SYS_ONOFF_H_
+
+#include <kernel.h>
+#include <zephyr/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup resource_apis Resource Reservation APIs
+ * @ingroup kernel_apis
+ * @{
+ */
+
+/**
+ * @brief Flag fields used to specify on-off service behavior.
+ */
+enum onoff_service_flags {
+	/**
+	 * @brief Flag passed to onoff_service_init().
+	 *
+	 * When provided this indicates the start transition function
+	 * is safe to invoke from ISR context.
+	 */
+	ONOFF_SERVICE_ISR_CALLABLE_START        = BIT(0),
+
+	/**
+	 * @brief Flag passed to onoff_service_init().
+	 *
+	 * When provided this indicates the start transition function
+	 * may suspend.
+	 */
+	ONOFF_SERVICE_SUSPENDABLE_START         = BIT(1),
```

Comment:
The intent of these flags is to specify transition function behavior as discussed in #20959, which is being actively revised; `suspend` is currently `rescheduling` or maybe `cooperative-safe`.

The transition functions may be invoked from thread or interrupt context.  At the moment there's a distinction between `isr-callable` and what is now `reschduling`.  The service can know whether it's being run from an ISR or a thread.  The flags are used to tell the service what the behavior of the transition function is, since it can't look at documentation, and from that determine a request can be satisfied.

At the moment there is only one code path that differs in behavior between the features, and it probably shouldn't.

This may get resolved when we settle on the terms.  It would be convenient if there was only one flag.